In [21]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

from sklearn.linear_model import SGDRegressor
from sklearn.model_selection import GridSearchCV

from sklearn.preprocessing import PolynomialFeatures

from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

from sklearn.model_selection import GridSearchCV

from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score

In [7]:
file_path = r"C:\Users\zeina\OneDrive\Desktop\New folder (3)\winequality-red.csv"
df = pd.read_csv(file_path)

In [8]:
#first few rows
print("First five rows:")
print(df.head())

#DataFrame information to understand data types
print("\nDataFrame info:")
print(df.info())

# summary statistics to see the data distributions
print("\nSummary statistics:")
print(df.describe())

First five rows:
   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2  

In [9]:
#removing missing values if there is any
missing_values = df.isnull().sum()
print("\nMissing values per column:")
print(missing_values)


Missing values per column:
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64


In [10]:
#Removing duplicate rows if there is any
duplicate_count = df.duplicated().sum()
print("\nNumber of duplicate rows:", duplicate_count)
df = df.drop_duplicates()


Number of duplicate rows: 240


In [11]:
print("Number of duplicate rows after removal:", df.duplicated().sum())

Number of duplicate rows after removal: 0


In [12]:
#list columns
print("Column names:")
print(df.columns)

Column names:
Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')


In [13]:
# Create feature matrix X by dropping the 'quality' column
X = df.drop(columns=['quality'])

# Create target vector y as the 'quality' column
y = df['quality']

# Verify by printing the first few rows
print("Features (X):")
print(X.head())

print("\nLabel (y):")
print(y.head())


Features (X):
   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
5            7.4              0.66         0.00             1.8      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
5                 13.0                  40.0   0.9978  3.51       0.56   

   alcohol  
0      9.4  
1      9.8  
2      9.8  
3      9.8  
5      

In [14]:
# Split the data: 80% for training, 20% for testing.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [15]:
# Display the shapes of the split data
print("Training Features shape:", X_train.shape)
print("Test Features shape:", X_test.shape)
print("Training Labels shape:", y_train.shape)
print("Test Labels shape:", y_test.shape)

Training Features shape: (1087, 11)
Test Features shape: (272, 11)
Training Labels shape: (1087,)
Test Labels shape: (272,)


In [16]:
# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training features and transform the training features
X_train_scaled = scaler.fit_transform(X_train)

# Use the same scaler to transform the test features
X_test_scaled = scaler.transform(X_test)

print("\nFirst few rows of scaled training features:")
display(pd.DataFrame(X_train_scaled, columns=X_train.columns).head())



First few rows of scaled training features:


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0       0.351623         -0.828323     0.648771       -0.238440  -0.411606   
1      -0.981548          0.965891    -0.989833        0.054589   0.312800   
2      -0.865620          0.180922    -1.245865       -0.604727  -0.106593   
3      -0.517836          0.853753    -0.580182       -0.531470  -0.220973   
4      -0.285980          0.349130    -0.477769       -0.165183  -0.259100   

   free sulfur dioxide  total sulfur dioxide   density        pH  sulphates  \
0            -0.362070             -0.566298 -0.986612 -0.259822   0.519851   
1            -0.168479             -0.596342 -0.025508  0.857267  -0.458489   
2             0.218703             -0.355992  0.055032  0.857267  -0.170742   
3            -0.362070             -0.235817  0.538268  1.251534   0.232104   
4             0.121907              0.214839  0.430882  0.791556  -0.400940   

    alcohol  
0  1.984240  
1 -0.215867  
2 -0.399209  
3 -0.490880  
4 -0.399209

## 1. Plain linear regression modeling

Training the model and evaluating its performance using Mean Squared Error (MSE), Root Mean Squared Error (RMSE), and R²

In [38]:
# Initialize the Linear Regression model
lin_reg = LinearRegression()

# Train the model on the scaled training data
lin_reg.fit(X_train_scaled, y_train)


LinearRegression()

In [39]:
# Predict on the training set
y_train_pred = lin_reg.predict(X_train_scaled)

# Predict on the test set
y_test_pred = lin_reg.predict(X_test_scaled)


Mean Squared Error (MSE): The average squared difference between actual and predicted values.

Root Mean Squared Error (RMSE): The square root of the MSE, providing an error value in the same 
units as the target.

R² (Coefficient of Determination): Represents the proportion of the variance in the dependent variable 
that is predictable from the independent variables.

In [ ]:
#performance on the training set
train_mse = mean_squared_error(y_train, y_train_pred)
train_rmse = np.sqrt(train_mse)
train_r2 = r2_score(y_train, y_train_pred)

#performance on the test set
test_mse = mean_squared_error(y_test, y_test_pred)
test_rmse = np.sqrt(test_mse)
test_r2 = r2_score(y_test, y_test_pred)

print("Training Set Metrics:")
print("  MSE: {:.3f}".format(train_mse))
print("  RMSE: {:.3f}".format(train_rmse))
print("  R²: {:.3f}".format(train_r2))
print("")
print("Test Set Metrics:")
print("  MSE: {:.3f}".format(test_mse))
print("  RMSE: {:.3f}".format(test_rmse))
print("  R²: {:.3f}".format(test_r2))

Training Set Metrics:
  MSE: 0.424
  RMSE: 0.651
  R²: 0.348

Test Set Metrics:
  MSE: 0.390
  RMSE: 0.625
  R²: 0.403


MSE: Subtraction of real wine quality from what the model predicts for each wine, then square those differences, and finally average them. That average is 0.424. Lower means better predictions.

RMSE: It tells us that on average, our model's prediction is about 0.65 quality points off from the true quality.

\( R^2 \): This number tells us that our model explains about 35% of the differences in wine quality. The remaining 65% of what influences the quality is not captured by the model.

## 2. SGDRegressor class and learning rate

The SGDRegressor class trains a linear model using stochastic gradient descent. 

In [ ]:
# Initialize the SGDRegressor with a constant learning rate (eta0=0.01 is chosen as an initial guess)
sgd_reg = SGDRegressor(max_iter=1000, tol=1e-3, learning_rate='constant', eta0=0.01, random_state=42)

# Train the model on the scaled training data
sgd_reg.fit(X_train_scaled, y_train)

#predictions on the training and test sets
y_train_pred_sgd = sgd_reg.predict(X_train_scaled)
y_test_pred_sgd = sgd_reg.predict(X_test_scaled)

# Compute evaluation metrics
train_mse_sgd = mean_squared_error(y_train, y_train_pred_sgd)
train_rmse_sgd = np.sqrt(train_mse_sgd)
train_r2_sgd = r2_score(y_train, y_train_pred_sgd)

test_mse_sgd = mean_squared_error(y_test, y_test_pred_sgd)
test_rmse_sgd = np.sqrt(test_mse_sgd)
test_r2_sgd = r2_score(y_test, y_test_pred_sgd)

print("SGDRegressor Training Set Metrics:")
print("  MSE: {:.3f}".format(train_mse_sgd))
print("  RMSE: {:.3f}".format(train_rmse_sgd))
print("  R²: {:.3f}".format(train_r2_sgd))
print("")
print("SGDRegressor Test Set Metrics:")
print("  MSE: {:.3f}".format(test_mse_sgd))
print("  RMSE: {:.3f}".format(test_rmse_sgd))
print("  R²: {:.3f}".format(test_r2_sgd))


SGDRegressor Training Set Metrics:
  MSE: 0.429
  RMSE: 0.655
  R²: 0.341

SGDRegressor Test Set Metrics:
  MSE: 0.400
  RMSE: 0.632
  R²: 0.388


### Finding the Correct Learning Rate

Tested several learning rates using cross-validation. I used GridSearchCV to search over a range of values for eta0.

In [ ]:
#parameter grid for the learning rate
param_grid = {
    'eta0': [0.0001, 0.001, 0.01, 0.1, 1.0],
    'learning_rate': ['constant', 'invscaling', 'adaptive']
}

# Initialize SGDRegressor
sgd = SGDRegressor(max_iter=1000, tol=1e-3, random_state=42)

# Setup GridSearchCV to search for the best learning rate parameters using 5-fold cross-validation.
grid_search = GridSearchCV(sgd, param_grid, cv=5, scoring='neg_mean_squared_error', verbose=1)
grid_search.fit(X_train_scaled, y_train)

# Get the best parameters and the corresponding score
best_params = grid_search.best_params_
best_score = np.sqrt(-grid_search.best_score_)

print("Best Parameters:", best_params)
print("Best cross-validated RMSE: {:.3f}".format(best_score))


Fitting 5 folds for each of 15 candidates, totalling 75 fits
Best Parameters: {'eta0': 0.01, 'learning_rate': 'invscaling'}
Best cross-validated RMSE: 0.663


## 3. Polynomial models

In [ ]:
# For demonstration, suppose we have a single feature or the whole X.
poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X_train_scaled)

print("Shape of original features:", X_train_scaled.shape)
print("Shape after polynomial transformation (degree=2):", X_poly.shape)


Shape of original features: (1279, 11)
Shape after polynomial transformation (degree=2): (1279, 77)


### Pipeline for Polynomial Regression

In [ ]:
#pipeline that scales the data, transforms it to polynomial features, then fits a linear regression model.
pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('poly', PolynomialFeatures()),  
    ('lin_reg', LinearRegression())
])


### Choosing the Best Polynomial Degree Using Grid Search

In [ ]:
#different polynomial degrees.
param_grid = {
    'poly__degree': [1, 2, 3, 4, 5]
}

# Setup grid search with 5-fold cross-validation using negative MSE as the scoring metric.
grid_search = GridSearchCV(pipeline, param_grid, cv=5, scoring='neg_mean_squared_error', verbose=1)
grid_search.fit(X_train_scaled, y_train)

# Retrieve the best polynomial degree and its associated performance.
best_degree = grid_search.best_params_['poly__degree']
best_rmse = np.sqrt(-grid_search.best_score_)

print("Best Polynomial Degree:", best_degree)
print("Best cross-validated RMSE: {:.3f}".format(best_rmse))


Fitting 5 folds for each of 5 candidates, totalling 25 fits
Best Polynomial Degree: 1
Best cross-validated RMSE: 0.663


### Training and Evaluating the Selected Polynomial Model with degree 1

In [48]:
# Retrieve the best model from the grid search
best_poly_model = grid_search.best_estimator_

# Predict on the test set
y_test_pred_poly = best_poly_model.predict(X_test_scaled)

# Evaluate performance on the test set
from sklearn.metrics import mean_squared_error, r2_score
test_mse_poly = mean_squared_error(y_test, y_test_pred_poly)
test_rmse_poly = np.sqrt(test_mse_poly)
test_r2_poly = r2_score(y_test, y_test_pred_poly)

print("Test Set Metrics for Polynomial Model (degree={}):".format(best_degree))
print("  MSE: {:.3f}".format(test_mse_poly))
print("  RMSE: {:.3f}".format(test_rmse_poly))
print("  R²: {:.3f}".format(test_r2_poly))


Test Set Metrics for Polynomial Model (degree=1):
  MSE: 0.390
  RMSE: 0.625
  R²: 0.403


**6. How We Utilized the Dataset**

- **Training Set:**  
  I used about 80% of the data to train our model, letting it learn the relationship between the features and wine quality.

- **Validation Set:**  
  Through cross-validation (e.g., 5-fold), I fine-tuned our model’s parameters, ensuring the model performs well on unseen parts of the training data.

- **Test Set:**  
  The remaining data was kept completely separate for the final evaluation, providing an unbiased measure of how well our model generalizes.


##  Lasso Model with Hyperparameter Tuning

Using scikit-learn’s Lasso class along with GridSearchCV to find the best alpha

In [ ]:

# Define a range of alpha values to search over.
alpha_range = [0.0001, 0.001, 0.01, 0.1, 1.0, 10.0]

# Create a parameter grid.
param_grid = {'alpha': alpha_range}

# Initialize Lasso with a high max_iter to ensure convergence.
lasso = Lasso(max_iter=10000, random_state=42)

# Setup GridSearchCV 
grid_search = GridSearchCV(lasso, param_grid, cv=5, scoring='neg_mean_squared_error', verbose=1)
grid_search.fit(X_train_scaled, y_train)

#best estimator
best_lasso = grid_search.best_estimator_
print("Best alpha: {:.4f}".format(best_lasso.alpha))

# Evaluate the model on the test set
y_test_pred_lasso = best_lasso.predict(X_test_scaled)
test_mse_lasso = mean_squared_error(y_test, y_test_pred_lasso)
test_rmse_lasso = np.sqrt(test_mse_lasso)
test_r2_lasso = r2_score(y_test, y_test_pred_lasso)

print("\nTest Set Metrics for Best Lasso Model:")
print("  MSE: {:.3f}".format(test_mse_lasso))
print("  RMSE: {:.3f}".format(test_rmse_lasso))
print("  R²: {:.3f}".format(test_r2_lasso))


Fitting 5 folds for each of 6 candidates, totalling 30 fits
Best alpha: 0.0100

Test Set Metrics for Best Lasso Model:
  MSE: 0.393
  RMSE: 0.627
  R²: 0.399


### Checking the Model's Coefficients

look at the model’s coefficients using the coef_ attribute to see the weight each feature gets.

In [50]:
# Create a DataFrame for easy inspection of feature names and coefficients.
coef_df = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': best_lasso.coef_
})

print("\nLasso Coefficients:")
display(coef_df)

# Identify and count features with coefficients practically zero.
# (We use a tolerance to capture values close to zero, e.g., 1e-3)
tolerance = 1e-3
filtered_features = coef_df[np.abs(coef_df['Coefficient']) < tolerance]
print("\nNumber of features filtered out (coefficients near zero):", filtered_features.shape[0])
print("Filtered features:")
display(filtered_features)



Lasso Coefficients:


Feature  Coefficient
0          fixed acidity     0.009898
1       volatile acidity    -0.171481
2            citric acid    -0.000000
3         residual sugar     0.000000
4              chlorides    -0.077486
5    free sulfur dioxide     0.030957
6   total sulfur dioxide    -0.095415
7                density    -0.000000
8                     pH    -0.048951
9              sulphates     0.131292
10               alcohol     0.297999


Number of features filtered out (coefficients near zero): 3
Filtered features:


Feature  Coefficient
2     citric acid         -0.0
3  residual sugar          0.0
7         density         -0.0

some features ended up with coefficients very close to or exactly zero, meaning Lasso effectively filtered them out.

## Model Comparison

### Plain Linear Regression

- **Simple Start:**  
  Basic model to get an initial understanding of the data.

- **Performance:**  
  - **RMSE:** ~0.625 (average prediction error of about 0.63 points)  
  - **R²:** ~0.403 (explains about 40% of the variance in wine quality)

- **Limitation:**  
  May not capture complex, non-linear relationships.

#

### Stochastic Gradient Descent (SGDRegressor)

- **Efficient:**  
  Handles large datasets well; performance can match plain linear regression when tuned correctly.

- **Tuning Sensitive:**  
  Requires careful selection of hyperparameters (e.g., learning rate via grid search).

- **Observation:**  
  With proper tuning, the model’s metrics are comparable to the baseline.


### Polynomial Regression

- **Captures Non-Linearity:**  
  Transforms features into higher-order terms to model complex patterns.

- **Degree Trade-Off:**  
  The key challenge here was selecting the optimal polynomial degree. A degree too low may underfit, while too high a degree can lead to overfitting.

- **Optimization:**  
  Cross-validation (e.g., testing degrees 1–5, typically degree 2 or 3 works best) helps improve RMSE and R².



### Lasso Regularization

- **Feature Filtering:**  
  Uses an L1 penalty to shrink unimportant feature coefficients to zero.

- **Parameter Control:**  
  The regularization strength (alpha) is tuned via grid search.

- **Benefits:**  
  Simplifies the model by filtering out irrelevant features, reduces overfitting, and improves interpretability (even if accuracy is marginally lower).


  ### Final Choice

- **If you need simplicity and clear results:**  
  Lasso is great because it reduces unimportant features and keeps the model simple.

- **If you want to capture complex, non-linear patterns:**  
  A well-tuned polynomial model may predict better.

- **For comparison:**  
  A basic linear model or an SGD model can serve as useful benchmarks.



**This assignment is completed individually.** 
